pymc3/distributions/continuous.py

```diff
@@ -632,101 +655,55 @@ class TruncatedNormal(BoundedContinuous):
 
     """
 
-    def __init__(
-        self,
-        mu=0,
+    rv_op = truncated_normal
+    bound_args_indices = (2, 3)  # indexes for lower and upper args
+
+    @classmethod
+    def dist(
+        cls,
+        mu=None,
         sigma=None,
         tau=None,
+        sd=None,
         lower=None,
         upper=None,
         transform="auto",
-        sd=None,
         *args,
         **kwargs,
     ):
-        if sd is not None:
-            sigma = sd
+        sigma = sd if sd is not None else sigma
         tau, sigma = get_tau_sigma(tau=tau, sigma=sigma)
-        self.sigma = self.sd = at.as_tensor_variable(sigma)
-        self.tau = at.as_tensor_variable(tau)
-        self.lower_check = at.as_tensor_variable(floatX(lower)) if lower is not None else lower
-        self.upper_check = at.as_tensor_variable(floatX(upper)) if upper is not None else upper
-        self.lower = (
+        sigma = at.as_tensor_variable(sigma)
+        tau = at.as_tensor_variable(tau)
+        mu = at.as_tensor_variable(floatX(mu))
+        assert_negative_support(sigma, "sigma", "TruncatedNormal")
+        assert_negative_support(tau, "tau", "TruncatedNormal")
+
+        if lower is None and upper is None:
+            initval = mu
+        elif lower is None and upper is not None:
+            initval = upper - 1.0
+        elif lower is not None and upper is None:
+            initval = lower + 1.0
+        else:
+            initval = (lower + upper) / 2
+
+        lower = (
```

Comment:
We shouldn't transform away from `None` unless it is needed, otherwise the branch logic in the logp (and also in the interval transform) is in vain, as it relies on a `None` check. Was this giving you problems.

In the `rng_fn` we can add extra logic to check for `None` and convert to `infs` if needed. If anything in the `dist` method we could check for `np.inf` and convert those to `None`.

Reply:
Sorry for that. I still have a partial understanding of the whole picture. Haven't had a chance yet to go through the sampling/inference part. So, I still don't know how `logp` is actually used (ie as understood takes as arguments what has already been processed by `dist`, ie the RandomVariable node inputs).

If the Nones are not handled `make_node` in `RandomVariable` would fail, as it's expecting objects that could be converted to tensor variables
https://github.com/aesara-devs/aesara/blob/3acdd78e7a5d27e70cd8de3fc00bd5f010c65ee3/aesara/tensor/random/op.py#L360

Stack trace
```
self = <pymc3.tests.test_distributions_random.TestTruncatedNormalLowerTau object at 0x7fb50c50a3d0>

    def test_distribution(self):
        self.validate_tests_list()
>       self._instantiate_pymc_rv()

pymc3/tests/test_distributions_random.py:288: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pymc3/tests/test_distributions_random.py:298: in _instantiate_pymc_rv
    self.pymc_rv = self.pymc_dist.dist(
pymc3/distributions/continuous.py:709: in dist
    res = super().dist([mu, sigma, lower, upper], testval=initval, **kwargs)
pymc3/distributions/distribution.py:285: in dist
    rv_out = cls.rv_op(*dist_params, size=create_size, **kwargs)
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py:323: in __call__
    res = super().__call__(rng, size, dtype, *args, **kwargs)
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/graph/op.py:271: in __call__
    node = self.make_node(*inputs, **kwargs)
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py:360: in make_node
    dist_params = tuple(
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py:361: in <genexpr>
    as_tensor_variable(p) if not isinstance(p, Variable) else p
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/__init__.py:41: in as_tensor_variable
    return _as_tensor_variable(x, name, ndim, **kwargs)
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/functools.py:875: in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
```

Reply:
I wonder whether checking for `inf` and `-inf` rather than `None` in `logp` and in the interval transform could be a feasible approach

Reply:
That's spot on. I will give it some thought.

One ugly solution is to have 3 versions of the distribution (each with its own transform and logp / random methods) but that feels super ugly.

Another is to wrap the infinity arguments in an `at.constant` which we can safely query (I think) when calling the logp method (first checking if it's a constant and then if it's +- infinity). 

For the interval transform we can define the different variations at creation time in the `dist` classmethod (this is done by the ...)

Reply:
Seems like you got ahead of me!

Reply:
> I wonder whether checking for `inf` and `-inf` rather than `None` in `logp` and in the interval transform could be a feasible approach

Yeah we can but only within an `at.constant` otherwise we would need symbolic branching with `at.switch` instead of the python `if else`. Since the point is to return a more optimized graph I would go with the constant + python branching.

Reply:
Do we have tests that cover this specific `+-inf` behavior?

Reply:
`TestTruncatedNormalLowerTau` and `TestTruncatedNormalUpperTau`

Reply:
I mean testing the interval and the logp is actually doing what it should (returning different expressions based on whether one or both of the inputs are `None`)

Reply:
I added `pymc3.tests.test_distributions.TestBoundedContinuous` to test the interval. I haven't been able to test `BoundedContinuous` directly, so I had to do so through a `TruncatedNormal`

I also added a slightly more convoluted test (which seems sensible) for the interval behavior `pymc3.tests.test_distributions.TestBoundedContinuous.test_missing_upper_bound_with_richer_context` which is failing. I need to investigate why

I added a couple of tests for the logp with unbounded intervals `pymc3.tests.test_distributions.TestMatchesScipy.test_truncated_normal`.

Let me know if you think it makes sense to add more